packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m

```diff
@@ -1313,9 +1314,45 @@ - (BOOL)setupWriterForPath:(NSString *)path {
   return YES;
 }
 
-- (void)setUpCaptureSessionForAudio {
+// this same function is also in video_player_avfoundation
+// configure application wide audio session manually to prevent overwriting
+// flag MixWithOthers by capture session, only change category if it is considered
+// as upgrade which means it can only enable ability to play in silent mode or
+// ability to record audio but never disables it, that could affect other plugins
+// which depend on this global state, only change category or options if there is
+// change to prevent unnecessary lags and silence
+// https://github.com/flutter/flutter/issues/131553
+static void upgradeAudioSessionCategory(AVAudioSessionCategory category,
+                                        AVAudioSessionCategoryOptions options,
+                                        AVAudioSessionCategoryOptions clearOptions) {
+  if (category == nil) {
+    category = AVAudioSession.sharedInstance.category;
+  }
+  NSSet *playCategories = [NSSet
+      setWithObjects:AVAudioSessionCategoryPlayback, AVAudioSessionCategoryPlayAndRecord, nil];
+  NSSet *recordCategories =
+      [NSSet setWithObjects:AVAudioSessionCategoryRecord, AVAudioSessionCategoryPlayAndRecord, nil];
+  NSSet *categories = [NSSet setWithObjects:category, AVAudioSession.sharedInstance.category, nil];
```

Comment:
what does this `categories` mean? why are we combining the set? Can you add some comments to explain?

Reply:
Categories which should be combined. Upgrading means we need to combine categories so the new category has play flavour if one of them had it, and has record when one of them had it. Test whether any of these categories has play is done by intersection which means if anything in the first set is present the second set.

Reply:
Can you rename this to `requiredFinalCategories` or something similar